### PR TITLE
Unit Tests Infrastructure Improvements

### PR DESCRIPTION
**Pull request by AlexanderSher**
This PR contains 2 changes:
- Makes log methods sync (product code change)
- Allows unit test fixtures to reference each other and assembly-level fixtures to create method-level fixtures (test code change)

### CODE REVIEW

File: src/Common/Core/Impl/Disposables/Disposable.cs [hunk 1]
```diff
@@ -25,6 +25,20 @@ public static class Disposable {
         }
 
         /// <summary>
+        /// Creates a disposable wrapper for a disposable that will be invoked at most once.
+        /// </summary>
+        /// <param name="disposable">Disposable that will be wrapped. Disposable is guaranteed to be run at most once.</param>
+        /// <returns>The disposable object that disposes wrapped object.</returns>
+        /// <exception cref="T:System.ArgumentNullException"><paramref name="disposable" /> is null.</exception>
+        public static IDisposable Create(IDisposable disposable) {
+            if (disposable == null) {
```
Comment (MikhailArkhipov) [inline src/Common/Core/Impl/Disposables/Disposable.cs]: You can use `Check.ArgumentNull(nameof(disposable), disposable);` for brevity if you wish

Comment (AlexanderSher) [inline src/Common/Core/Impl/Disposables/Disposable.cs]: Sure!

File: src/Common/Core/Test/Logging/LoggerTest.cs [hunk 2]
```diff
@@ -18,30 +17,35 @@ public class LoggerTest {
...
             int i = 0;
-            foreach(var v in Enum.GetValues(typeof(LogVerbosity))) {
-                if ((int)v > (int)LogVerbosity.None && (int)v <= (int)verbosity) {
-                    await writer.Received().WriteAsync(MessageCategory.Error, "message" + i.ToString());
-                } else {
-                    await writer.DidNotReceive().WriteAsync(MessageCategory.Error, "message" + i.ToString());
+            foreach (var v in Enum.GetValues(typeof(LogVerbosity)))
+            {
+                if ((int)v > (int)LogVerbosity.None && (int)v <= (int)verbosity)
+                {
```
Comment (MikhailArkhipov) [inline src/Common/Core/Test/Logging/LoggerTest.cs]: Formatting

File: src/Host/Client/Impl/Host/BrokerClient.cs [hunk 3]
```diff
@@ -139,60 +139,65 @@ internal abstract class BrokerClient : IBrokerClient {
...
-            var sessions = new SessionsWebService(HttpClient, _credentials);
-            try {
-                await sessions.PutAsync(name, new SessionCreateRequest {
-                    InterpreterId = _interpreterId,
-                    CommandLineArguments = rCommandLineArguments,
-                }, cancellationToken);
-            } catch (BrokerApiErrorException apiex) {
-                throw new RHostDisconnectedException(MessageFromBrokerApiException(apiex), apiex);
+            var sessions = new SessionsWebService(HttpClient, _credentials, Log);
+            using (Log.Measure(LogVerbosity.Normal, $"Create broker session \"{name}\"")) {
```
Comment (MikhailArkhipov) [inline src/Host/Client/Impl/Host/BrokerClient.cs]: Evil FxCop will want Invariant

File: src/Host/Client/Impl/Host/BrokerClient.cs [hunk 4]
```diff
@@ -139,60 +139,65 @@ internal abstract class BrokerClient : IBrokerClient {
...
-                InspectResponse = response => {
-                    if (response.StatusCode == HttpStatusCode.Forbidden) {
-                        throw new UnauthorizedAccessException();
+            using (Log.Measure(LogVerbosity.Normal, $"Connect to broker session \"{name}\"")) {
+                var wsClient = new WebSocketClient {
+                    KeepAliveInterval = HeartbeatTimeout,
+                    SubProtocols = { "Microsoft.R.Host" },
+                    InspectResponse = response => {
+                        if (response.StatusCode == HttpStatusCode.Forbidden)
+                        {
```
Comment (MikhailArkhipov) [inline src/Host/Client/Impl/Host/BrokerClient.cs]: Formatting

File: src/Host/Client/Impl/Host/BrokerClient.cs [hunk 5]
```diff
@@ -139,60 +139,65 @@ internal abstract class BrokerClient : IBrokerClient {
...
-                using (await _credentials.LockCredentialsAsync(cancellationToken)) {
-                    try {
-                        request.AuthenticationLevel = AuthenticationLevel.MutualAuthRequested;
-                        request.Credentials = HttpClientHandler.Credentials;
-                        return await wsClient.ConnectAsync(request, cancellationToken);
-                    } catch (UnauthorizedAccessException) {
-                        _credentials.InvalidateCredentials();
-                    } catch (Exception ex) when (ex is InvalidOperationException) {
-                        throw new RHostDisconnectedException(Resources.HttpErrorCreatingSession.FormatInvariant(Name, ex.Message), ex);
+                var pipeUri = new UriBuilder(HttpClient.BaseAddress) {
```
Comment (MikhailArkhipov) [inline src/Host/Client/Impl/Host/BrokerClient.cs]: Not sure what is exactly different here, fragments appear the same...

Comment (MikhailArkhipov) [inline src/Host/Client/Impl/Host/BrokerClient.cs]: Whitespace?

File: src/Host/Client/Impl/Session/RSession.cs [hunk 6]
```diff
@@ -99,6 +100,7 @@ internal sealed class RSession : IRSession, IRCallbacks {
             });
 
             _initializationLock = initializationLock;
+            _log = log;
```
Comment (MikhailArkhipov) [inline src/Host/Client/Impl/Session/RSession.cs]: Is it used anywhere? It seems to be using ` _host?.Log` instead

**Comment by MikhailArkhipov** [comment]
👍

File: src/Host/Client/Impl/Session/RSession.cs (hunk 6 above)
Comment (AlexanderSher) [inline src/Host/Client/Impl/Session/RSession.cs]: Yeah, good catch, we don't use this field.

File: src/Host/Client/Impl/Host/BrokerClient.cs (hunk 3 above)
Comment (AlexanderSher) [inline src/Host/Client/Impl/Host/BrokerClient.cs]: Fixed

**Pull request merged**